### PR TITLE
[Breaking change] Simplify object store heap size configuration

### PR DESCRIPTION
**Pull request by yrajas**
This change simplifies configuring object store heap size for continuous monitoring. Instead of configuring total object store size, memory constraints for object store can now be configured using the following simple settings.

obj-index - Size of Object store index
obj-log-memory - Size of object store log that contains references to heap objects (Previously available as --obj-memory)
obj-heap-memory - Size of object store heap

Previously available --obj-total-memory option is now removed as object store memory can now be controlled using the above options in a fine grained manner.

Total object store memory footprint = obj-index + obj-log-memory + obj-heap-memory.
No changes to how the main store size is configured.

- [x] Simplified object store memory configuration
- [x] Added additional details on GC memory to INFO MEMORY command output. This also includes memory that is held by GC not yet released back to the system.
- [x] Bug fix to account for null IGarnetObject.
- [x] Updated documentation.

### CODE REVIEW

File: libs/host/Configuration/Options.cs [hunk 1]
```diff
@@ -64,11 +64,11 @@ internal sealed class Options
         public int MutablePercent { get; set; }
 
         [MemorySizeValidation(false)]
-        [Option("obj-total-memory", Required = false, HelpText = "Total object store log memory used including heap memory in bytes")]
-        public string ObjectStoreTotalMemorySize { get; set; }
+        [Option("obj-heap-memory", Required = false, HelpText = "Object store heap memory size in bytes")]
```
Comment (badrishc) [inline libs/host/Configuration/Options.cs:67]: explain what heap memory means - "Object store heap memory size (to account for the size taken up by object instances on the heap)"

File: libs/server/Metrics/Info/GarnetInfoMetrics.cs [hunk 2]
```diff
@@ -82,6 +82,9 @@ private void PopulateMemoryInfo(StoreWrapper storeWrapper)
                 total_object_store_size = object_store_index_size + object_store_log_memory_references_size + object_store_read_cache_size;
             }
 
+            var gcMemoryInfo = GC.GetGCMemoryInfo();
```
Comment (badrishc) [inline libs/server/Metrics/Info/GarnetInfoMetrics.cs:85]: add object_store_heap_memory_size and total include that as well.

File: libs/server/Metrics/Info/GarnetInfoMetrics.cs [hunk 3]
```diff
@@ -105,14 +108,17 @@ private void PopulateMemoryInfo(StoreWrapper storeWrapper)
...
                 new("proc_peak_physical_memory_size", SystemMetrics.GetPeakPhysicalMemoryUsage().ToString()),
                 new("proc_peak_physical_memory_size(MB)", SystemMetrics.GetPeakPhysicalMemoryUsage(1 << 20).ToString()),
+                new("gc_heap_bytes", gcMemoryInfo.HeapSizeBytes.ToString()),
+                new("gc_managed_memory_bytes_excluding_heap", gcAvailableMemory.ToString()),
+                new("gc_fragmented_bytes", gcMemoryInfo.FragmentedBytes.ToString()),
                 new("main_store_index_size", main_store_index_size.ToString()),
                 new("main_store_log_memory_size", main_store_log_memory_size.ToString()),
                 new("main_store_read_cache_size", main_store_read_cache_size.ToString()),
                 new("total_main_store_size", total_main_store_size.ToString()),
                 new("object_store_index_size", object_store_index_size.ToString()),
```
Comment (badrishc) [inline libs/server/Metrics/Info/GarnetInfoMetrics.cs:118]: these names and the names in options should be made consistent

File: website/docs/getting-started/configuration.md [hunk 4]
```diff
@@ -77,8 +77,8 @@ For all available command line settings, run `GarnetServer.exe -h` or `GarnetSer
 | **IndexSize** | ```-i```<br/>```--index``` | ```string``` | Memory size | Size of hash index in bytes (rounds down to power of 2) |
 | **IndexMaxSize** | ```--index-max-size``` | ```string``` | Memory size | Max size of hash index in bytes (rounds down to power of 2) |
 | **MutablePercent** | ```--mutable-percent``` | ```int``` |  | Percentage of log memory that is kept mutable |
-| **ObjectStoreTotalMemorySize** | ```--obj-total-memory``` | ```string``` | Memory size | Total object store log memory used including heap memory in bytes |
-| **ObjectStoreLogMemorySize** | ```--obj-memory``` | ```string``` | Memory size | Object store log memory used in bytes excluding heap memory |
+| **ObjectStoreHeapMemorySize** | ```--obj-heap-memory``` | ```string``` | Memory size | Object store heap memory size in bytes |
+| **ObjectStoreLogMemorySize** | ```--obj-log-memory``` | ```string``` | Memory size | Object store log memory used in bytes excluding heap memory |
```
Comment (badrishc) [inline website/docs/getting-started/configuration.md:81]: Object store log memory used in bytes (only the log with references, excluding heap memory of objects pointed to from the log)

File: libs/host/Configuration/Options.cs [hunk 5]
```diff
@@ -64,11 +64,11 @@ internal sealed class Options
...
+        public string ObjectStoreHeapMemorySize { get; set; }
 
         [MemorySizeValidation]
-        [Option("obj-memory", Required = false, HelpText = "Object store log memory used in bytes excluding heap memory")]
+        [Option("obj-log-memory", Required = false, HelpText = "Object store log memory used in bytes excluding heap memory")]
```
Comment (badrishc) [inline libs/host/Configuration/Options.cs:71]: Object store log memory used in bytes (only the log with references, excluding heap memory of objects pointed to from the log)

**Review by badrishc — CHANGES_REQUESTED** [review]

File: libs/server/Metrics/Info/GarnetInfoMetrics.cs [hunk 6]
```diff
@@ -69,19 +69,26 @@ private void PopulateMemoryInfo(StoreWrapper storeWrapper)
...
             var disableObj = storeWrapper.serverOptions.DisableObjects;
 
+            var aof_log_memory_size = storeWrapper.appendOnlyFile?.MemorySizeBytes ?? -1;
+
             if (!disableObj)
             {
                 object_store_index_size = storeWrapper.objectStore.IndexSize * 64;
-                object_store_log_memory_references_size = storeWrapper.objectStore.Log.MemorySizeBytes;
+                object_store_log_memory_size = storeWrapper.objectStore.Log.MemorySizeBytes;
                 object_store_read_cache_size = (storeWrapper.objectStore.ReadCache != null ? storeWrapper.objectStore.ReadCache.MemorySizeBytes : 0);
```
Comment (badrishc) [inline libs/server/Metrics/Info/GarnetInfoMetrics.cs:84]: why does read_cache not have heap_memory and log_memory split out?

File: libs/server/Storage/SizeTracker/CacheSizeTracker.cs [hunk 7]
```diff
@@ -96,13 +95,8 @@ public void Start(CancellationToken token)
...
         public (long mainLogSizeBytes, long readCacheSizeBytes) CalculateLogTargetSizeBytes(long newTargetSize)
         {
-            long residual = newTargetSize - IndexSizeBytes;
-
-            if (residual <= 0)
-                throw new TsavoriteException($"Target size {newTargetSize} must be larger than index size {IndexSizeBytes}");
-
-            var mainLogSizeBytes = this.store.ReadCache == null ? residual : residual / 2;
-            var readCacheSizeBytes = this.store.ReadCache == null ? 0 : residual / 2;
+            var mainLogSizeBytes = this.store.ReadCache == null ? newTargetSize : newTargetSize / 2;
```
Comment (badrishc) [inline libs/server/Storage/SizeTracker/CacheSizeTracker.cs:98]: there is an implicit assumption that we split main log and read cache 50:50 here. this needs to be configured by user

File: website/docs/getting-started/configuration.md [hunk 8]
```diff
@@ -77,8 +77,8 @@ For all available command line settings, run `GarnetServer.exe -h` or `GarnetSer
...
-| **ObjectStoreLogMemorySize** | ```--obj-memory``` | ```string``` | Memory size | Object store log memory used in bytes excluding heap memory |
+| **ObjectStoreHeapMemorySize** | ```--obj-heap-memory``` | ```string``` | Memory size | Object store heap memory size in bytes (Sum of size taken up by all object instances in the heap) |
+| **ObjectStoreLogMemorySize** | ```--obj-log-memory``` | ```string``` | Memory size | Object store log memory used in bytes (Size of only the log with references to heap objects, excludes size of heap memory consumed by the objects themselves referred to from the log) |
 | **ObjectStorePageSize** | ```--obj-page``` | ```string``` | Memory size | Size of each object store page in bytes (rounds down to power of 2) |
 | **ObjectStoreSegmentSize** | ```--obj-segment``` | ```string``` | Memory size | Size of each object store log segment in bytes on disk (rounds down to power of 2) |
 | **ObjectStoreIndexSize** | ```--obj-index``` | ```string``` | Memory size | Size of object store hash index in bytes (rounds down to power of 2) |
```
Comment (badrishc) [inline website/docs/getting-started/configuration.md:84]: should this be IndexStartSize and ObjectStoreIndexStartSize

File: website/docs/getting-started/configuration.md [hunk 9]
```diff
@@ -77,8 +77,8 @@ For all available command line settings, run `GarnetServer.exe -h` or `GarnetSer
 | **IndexSize** | ```-i```<br/>```--index``` | ```string``` | Memory size | Size of hash index in bytes (rounds down to power of 2) |
 | **IndexMaxSize** | ```--index-max-size``` | ```string``` | Memory size | Max size of hash index in bytes (rounds down to power of 2) |
 | **MutablePercent** | ```--mutable-percent``` | ```int``` |  | Percentage of log memory that is kept mutable |
-| **ObjectStoreTotalMemorySize** | ```--obj-total-memory``` | ```string``` | Memory size | Total object store log memory used including heap memory in bytes |
-| **ObjectStoreLogMemorySize** | ```--obj-memory``` | ```string``` | Memory size | Object store log memory used in bytes excluding heap memory |
+| **ObjectStoreHeapMemorySize** | ```--obj-heap-memory``` | ```string``` | Memory size | Object store heap memory size in bytes (Sum of size taken up by all object instances in the heap) |
+| **ObjectStoreLogMemorySize** | ```--obj-log-memory``` | ```string``` | Memory size | Object store log memory used in bytes (Size of only the log with references to heap objects, excludes size of heap memory consumed by the objects themselves referred to from the log) |
```
Comment (badrishc) [inline website/docs/getting-started/configuration.md:81]: Also, how does user configure the read cache sizes for main and object?

File: libs/server/Storage/SizeTracker/CacheSizeTracker.cs (hunk 7 above)
Comment (yrajas) [reply libs/server/Storage/SizeTracker/CacheSizeTracker.cs:98]: These could change in a separate PR that enables read cache in Garnet.

File: website/docs/getting-started/configuration.md (hunk 9 above)
Comment (yrajas) [reply website/docs/getting-started/configuration.md:81]: Enabling read cache and its configuration could be part of a separate PR.

**Review by badrishc — APPROVED** [review]

**Pull request merged**